ansible.go

```diff
@@ -4,65 +4,105 @@ package main
 	"os"
 	"path/filepath"
 	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
 )
 
 // AnsibleConfig is a collection of meta-information about an Ansible repository.
 //
 // The virtualenv specified by the VenvConfig needs to be initialized before running Ansible commands.
 // All dir paths are relative to the root of the source tarball.
 type AnsibleConfig struct {
-	VenvConfig         VenvConfig // Virtualenv config that Ansible will be executed in
-	Cwd                string     // Path to change to when running Ansible commands
-	InventoryList      []string   // Paths to all desired inventories
+	VenvConfig    VenvConfig // Virtualenv config that Ansible will be executed in
+	Cwd           string     // Path to change to when running Ansible commands
+	InventoryList []string   // Paths to all desired inventories
+}
+
+// CreateAnsibleTargetsList generates and returns an array of possible targets
+// The possible targets are ip addresse from the host interfaces or the hostname itself
+func CreateAnsibleTargetsList() []string {
+	var inventoryTargets = []string{}
+
+	// extract all ip addresses except the loopback one and add the hostname as a last fallback mechanism
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		logrus.Fatal("Unable to get the network interfaces")
+	}
+	for _, i := range ifaces {
+		addrs, err := i.Addrs()
+		if err != nil {
+			logrus.Fatal("Unable to extract the ip addresses")
+		}
+		for _, addr := range addrs {
+			var ip net.IP
+			switch v := addr.(type) {
+			case *net.IPNet:
+				ip = v.IP
+			case *net.IPAddr:
+				ip = v.IP
+			}
+			if ip == nil || ip.IsLoopback() {
+				continue
+			}
+			inventoryTargets = append(inventoryTargets, ip.String())
+		}
+	}
+	inventoryTargets = append(inventoryTargets, hostname)
+	return inventoryTargets
 }
 
-// FindInventoryForHost takes a hostname and returns the inventory where the host was found.
+// FindInventoryForHost calls CreateAnsibleTargetsList to get a list of possible targets
+// and returns the traget and inventory where the target was found.
+// It stops on the first target match meaning that it assumes that hosts are
+// defined by only 1 type of target like their eth0 ip address or hostname
 // If the host was not found, it returns an error.
 //
 // This will check against all of the defined inventories in ansibleCfg.InventoryList,
 // relative to the path defined in ansibleCfg.Cwd.
 //
 // The given hostname should be the full name that appears in the Ansible inventories.
-func (a AnsibleConfig) FindInventoryForHost(host string) (string, error) {
-	for _, item := range a.InventoryList {
-		inv := filepath.Join(a.Cwd, item)
-		_, err := os.Stat(inv)
-		if err != nil {
-			if os.IsNotExist(err) {
-				return "", errors.Wrapf(err, "unable to find inventory: %s", item)
+func (a AnsibleConfig) FindInventoryForHost() (string, string, error) {
+	targets := CreateAnsibleTargetsList()
+	for _, target := range targets {
```

Comment:
So, it looks like there's a lot of setup that we do to get an inventory list out of ansible. Calling venv commands, etc.
It would probably be better if inverted these loops and pushed the targets list further down the function so we're not incessantly calling Ansible.

Reply:
PTAL